ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -454,12 +464,24 @@ private int numberOfEntries(ByteBuffer mmap) {
     return (mmap.capacity() - indexSizeExcludingEntries) / persistedEntrySize;
   }
 
-  private StoreKey getKeyAt(ByteBuffer mmap, int index) throws IOException {
-    mmap.position(firstKeyRelativeOffset + index * persistedEntrySize);
-    return factory.getStoreKey(new DataInputStream(new ByteBufferInputStream(mmap)));
+  private StoreKey getKeyAt(ByteBuffer mmap, int index) throws StoreException {
+    StoreKey storeKey = null;
+    try {
+      mmap.position(firstKeyRelativeOffset + index * persistedEntrySize);
+      storeKey = factory.getStoreKey(new DataInputStream(new ByteBufferInputStream(mmap)));
+    } catch (Throwable t) {
+      if (t.getClass() == InternalError.class && t.getMessage().equals(StoreException.INTERNAL_ERROR_STR)) {
+        throw new StoreException("Internal error occurred due to unsafe memory access", t, StoreErrorCodes.IOError);
+      } else if (t.getClass() == IOException.class && t.getMessage().equals(StoreException.IO_ERROR_STR)) {
+        throw new StoreException("IO error while trying to get store key", t, StoreErrorCodes.IOError);
```

Comment:
all of these exceptions with `StoreErrorCodes.IOError` can be counted by the `BlobStore` itself